core/connection.go

```diff
@@ -0,0 +1,77 @@
+package connection
+
+import (
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+type PermanentError struct {
+	Inner error
+}
+
+func (e PermanentError) Error() string { return e.Inner.Error() }
+func (e PermanentError) Unwrap() error {
+	return e.Inner
+}
+func (e PermanentError) Is(err error) bool {
+	_, ok := err.(PermanentError)
+	return ok
+}
+
+// Same as Retry only that the functionToRetry can return a value upon correct execution
+func RetryWithData[T any](functionToRetry func() (*T, error), minDelay uint64, factor float64, maxTries uint64) (*T, error) {
+	f := func() (*T, error) {
+		val, err := functionToRetry()
+		if perm, ok := err.(PermanentError); err != nil && ok {
```

Comment:
Type assertion on errors fails on wrapped errors

Reply:
good catch

Reply:
In fact, `err != nil` is redundant as `ok` already ensures that err is non-nil.